apypie/foreman.py

```diff
@@ -0,0 +1,239 @@
+                action, resource, str(exc))
+            raise ForemanApiException.from_exception(exc, msg) from exc
+        return result
+
+    def wait_for_task(self, task, ignore_errors=False):
+        """
+        Wait for a foreman-tasks task, polling it every ``self.task_poll`` seconds.
+
+        Will raise a ForemanApiException when task has not finished in ``self.task_timeout`` seconds.
+        """
+        duration = self.task_timeout
+        while task['state'] not in ['paused', 'stopped']:
+            duration -= self.task_poll
+            if duration <= 0:
+                raise ForemanApiException(msg="Timeout waiting for Task {0}".format(task['id']))
+            time.sleep(self.task_poll)
+
+            resource_payload = self._resource_prepare_params('foreman_tasks', 'show', {'id': task['id']})
+            task = self._resource_call('foreman_tasks', 'show', resource_payload)
+        if not ignore_errors and task['result'] != 'success':
+            msg = 'Task {0}({1}) did not succeed. Task information: {2}'.format(task['action'], task['id'], task['humanized']['errors'])
+            raise ForemanApiException(msg=msg)
+        return task
+
+    def show(self, resource, resource_id, params=None):
+        """
+        Execute the ``show`` action on an entity.
+
+        :param resource: Plural name of the api resource to show
+        :type resource: str
+        :param resource_id: The ID of the entity to show
+        :type resource_id: int
+        :param params: Lookup parameters (i.e. parent_id for nested entities)
+        :type params: Union[dict,None], optional
+        """
+        payload = {'id': resource_id}
+        if params:
+            payload.update(params)
+        return self.resource_action(resource, 'show', payload)
+
+    def list(self, resource, search=None, params=None):
+        """
+        Execute the ``index`` action on an resource.
+
+        :param resource: Plural name of the api resource to show
+        :type resource: str
+        :param search: Search string as accepted by the API to limit the results
+        :type search: str, optional
+        :param params: Lookup parameters (i.e. parent_id for nested entities)
+        :type params: Union[dict,None], optional
+        """
+        payload = {'per_page': PER_PAGE}
+        if search is not None:
+            payload['search'] = search
+        if params:
+            payload.update(params)
+
+        return self.resource_action(resource, 'index', payload)['results']
+
+    def create(self, resource, desired_entity, params=None):
+        """
+        Create entity with given properties
+
+        :param resource: Plural name of the api resource to manipulate
+        :type resource: str
+        :param desired_entity: Desired properties of the entity
+        :type desired_entity: dict
+        :param params: Lookup parameters (i.e. parent_id for nested entities)
+        :type params: dict, optional
```

Comment:
Why are some a union of dict and None while this is plain dict?

Reply:
Because the doc strings a copied from old code and need cleanup :)

Reply:
Now that we've dropped Python 2.7, should we move to real type annotations?

Reply:
Yes.

Reply:
done